pkg/ingester/circuitbreaker.go

```diff
@@ -0,0 +1,249 @@
+		WithDelay(cfg.CooldownPeriod).
+		OnClose(func(event circuitbreaker.StateChangedEvent) {
+			transitionClosed.Inc()
+			level.Info(logger).Log("msg", "circuit breaker is closed", "ingester", ingesterID, "previous", event.OldState, "current", event.NewState)
+		}).
+		OnOpen(func(event circuitbreaker.StateChangedEvent) {
+			transitionOpen.Inc()
+			level.Warn(logger).Log("msg", "circuit breaker is open", "ingester", ingesterID, "previous", event.OldState, "current", event.NewState)
+		}).
+		OnHalfOpen(func(event circuitbreaker.StateChangedEvent) {
+			transitionHalfOpen.Inc()
+			level.Info(logger).Log("msg", "circuit breaker is half-open", "ingester", ingesterID, "previous", event.OldState, "current", event.NewState)
+		}).
+		HandleIf(func(_ any, err error) bool { return isFailure(err) })
+
+	if cfg.testModeEnabled {
+		// In case of testing purposes, we initialize the circuit breaker with count based failure thresholding,
+		// since it is more deterministic, and therefore it is easier to predict the outcome.
+		cbBuilder = cbBuilder.WithFailureThreshold(cfg.FailureThreshold)
+	} else {
+		// In case of production code, we prefer time based failure thresholding.
+		cbBuilder = cbBuilder.WithFailureRateThreshold(cfg.FailureThreshold, cfg.FailureExecutionThreshold, cfg.ThresholdingPeriod)
+	}
+
+	cb.CircuitBreaker = cbBuilder.Build()
+	return &cb
+}
+
+func isFailure(err error) bool {
+	if err == nil {
+		return false
+	}
+
+	// We only consider timeouts or ingester hitting a per-instance limit
+	// to be errors worthy of tripping the circuit breaker since these
+	// are specific to a particular ingester, not a user or request.
+
+	if errors.Is(err, context.DeadlineExceeded) {
+		return true
+	}
+
+	statusCode := grpcutil.ErrorToStatusCode(err)
+	if statusCode == codes.DeadlineExceeded {
+		return true
+	}
+
+	var ingesterErr ingesterError
+	if errors.As(err, &ingesterErr) {
+		return ingesterErr.errorCause() == mimirpb.INSTANCE_LIMIT
+	}
+
+	return false
+}
+
+func (cb *circuitBreaker) State() circuitbreaker.State {
+	return cb.CircuitBreaker.State()
+}
+
+func (cb *circuitBreaker) isActive() bool {
+	if cb == nil {
+		return false
+	}
+	return cb.startTime.Before(time.Now())
+}
+
+func (cb *circuitBreaker) tryAcquirePermit() error {
+	if !cb.isActive() {
+		return nil
+	}
+	if !cb.CircuitBreaker.TryAcquirePermit() {
+		cb.metrics.circuitBreakerResults.WithLabelValues(cb.ingesterID, resultOpen).Inc()
+		return middleware.DoNotLogError{Err: newCircuitBreakerOpenError(cb.RemainingDelay())}
+	}
+	return nil
+}
+
+func (cb *circuitBreaker) recordSuccess() {
+	if !cb.isActive() {
+		return
+	}
+	cb.CircuitBreaker.RecordSuccess()
+	cb.metrics.circuitBreakerResults.WithLabelValues(cb.ingesterID, resultSuccess).Inc()
+}
+
+func (cb *circuitBreaker) recordError(err error) {
+	if !cb.isActive() {
+		return
+	}
+	cb.CircuitBreaker.RecordError(err)
+	cb.metrics.circuitBreakerResults.WithLabelValues(cb.ingesterID, resultError).Inc()
```

Comment:
here we always increase `error`, even if circuit breaker considers `err` to be a success

Reply:
Good observation. A better solution would be to use `CircuitBreakerBuilder.OnSuccess` and `OnFailure` for recording these metric values when a recorded result/error is considered a failure or not (per isFailure). In that case, the `recordSuccess` and `recordFailure` methods can probably be removed.

Reply:
I have implemented `recordResult()` method where I register an error and increment the error metrics only if the error represents an actual failure, and increment the success metrics otherwise.